Fix surface boundary conditions for TEOS-10

mjlosch opened the pull request:
## What changes does this PR introduce?
This a bug fix towards completing the TEOS-10 implementation


## What is the current behaviour? 
When using TEOS-10 (`'eosType='TEOS10'`, in `data&PARMS01`), the variable `THETA` is interpreted as conservative temperatures (CT), which introduces a bias at in bulk formulae at the surface. Therefore CT needs to be converted to potential (or at the surface: in-situ) temperature before using it in bulk formulae.

## What is the new behaviour 
add a new field `seaSurfTemp` that is computed once to be used in all bulk formulae. Convert this from CT to potential/in-situ temperature if 'eosType='TEOS10'`.

## Does this PR introduce a breaking change? 
- changes results when using `TEOS10`
- moved `sstExtrapol` from `pkg/exf` to the main namelist `data&PARMS01`, I doubt that anyone uses this.

## Other information:
This is a draft to leave room for discussion as I may have missed a few points.
Also, I moved the extrapolation parameter `sstExtrapol` from `pkg/exf` to the main namelist `data&PARMS01`, so that it is available to all code. We might want to discuss a new name for this or find a good way of retiring it from the namelist without changing the name.

## Suggested addition to `tag-index`
T.B.D.

jrscott commented:
@mjlosch in principal this looks great. Comments:

- can we confirm/clean up the somewhat sloppy language in manual 3.8.3 re TEOS (that I wrote; realize this extends a bit beyond this PR)
- sstExtrapol was presumably a nuisance issue for you here, I'm wondering too if we can put in a retired list (we can discuss). Note it was not a documented namelist parm in exf. Wouldn't usage be suspect w/TEOS?
-  not crazy about name seaSurfTemp because this is only used for pkgs exf and bulkformula (for reasons you outline) whereas theta(k=1) is common SST value. Might suggest changing name, or at least being specific about this in description in DYNVARS.h
- wondering if we should also have a z-coor TEOS10 verification run?

mjlosch commented:
- Let's do the documentation part in a different PR
- `sstExtrapol` is moved back to `pkg/exf` but now computation is now only once in `exf_getforcing.F`
- `seaSurfTemp` renamed to `surfLevTemp`
- @jm-c will think about an additional verification experiment.

mjlosch commented:
Note, that the merge changes the results of `global_ocean.cs32x15.in_p` and the reference needs to be re-generated.

jm-c commented on model/src/load_fields_driver.F:
```diff
@@ -127,6 +134,43 @@ SUBROUTINE LOAD_FIELDS_DRIVER( myTime, myIter, myThid )
+          DO j=1-OLy,sNy+OLy
+           DO i=1-OLx,sNx+OLx
+            surfLevTemp(i,j,bi,bj) =
+     &           surfLevTemp(i,j,bi,bj)*maskInC(i,j,bi,bj)
```
It might be safer to use "maskC(ks)" instead of "maskInC", a) if below an ice-shelf or b) if using OBCS.

jm-c reviewed (commented):
I have few comments:

1. definitions of constants in `EOS.h` and in `seawater.F` (s/r CONVERT_CT2PT) as "PARAMETER": previously, all the EOS constants were defined as variable in common block and set in `ini_eos.F`. Here the new ones are defined as PARAMETER. Is there a reason not to follow the same approach ? The other thing is that some are defined from arithmetic expression and I am not 100% sure that this works with any compiler. And finally, there are many "_d 0" missing. Just to check I printed the value of "I_cp0" and the value of:
 > tmpPrt = 1.0 _d 0/3991.86795711963 _d 0

    and it's different, plus the differences depend on compiler + processor.

2. The new variable "surfLevTemp" could be moved from `DYNVARS.h` to `FFIELDS.h`: this would not save the need to add `DYNVARS.h` in `load_fields_driver.F` or `exf_getforcing.F`, but could be removed from `exf_radiation.F` and initialisation will move from `ini_dynvars.F` to `ini_ffields.F`. I agree that it's not completely obvious which of these 2 header files is better, but the fact that the TAF storage dir for this variable has been added to `check_lev1_dir_forcing.h`, as it's used for surface forcing,   would lean towards FFIELDS.h (+ there are already few arrays there that are not strictly speaking "forcing" like relaxation masks).
3. the name `surfLevTemp` might not be the best, because it does not convey the fact that it's in-situ temperature. But since it has already been changed once, it should not be changed again unless everyone agree. Also, it might be good to mention that it does not follow the ocean surface level `kSurf` (or `kLowC` if P-coords) under ice-shelf ?
4. regarding new pkg/exf variable `exf_Tsf`, does it need to be in common block ? or just as a local var that would be passed as arg. to `exf_radiation.F` and `exf_bulkformulae.F` ?
5. I am getting few unused vars warning (4) from `exf_bulkformulae.F` & `exf_radiation.F`.

jm-c commented on model/src/seawater.F:
```diff
@@ -174,3 +182,235 @@ _RL FUNCTION SW_ADTG  (S,T,P)
+        Tp(i,j) = Tp_old(i,j) - Tc_diff(i,j)*dTp_dTc(i,j)
+C--   Estimate the potential temperature and its derivative from an C
+C     approximate rational function fit.
+        x2 = MAX(I_S0 * Sa(i,j), 0.0)
```
It's not safe to mix arg. of different type in intrinsic function, in thic case a _RL variable and a real*4 constant.
this is the main reason why "zeroRL", "zeroRS",  ... are there.

jm-c commented on pkg/exf/exf_getforcing.F:
```diff
@@ -144,9 +148,44 @@ SUBROUTINE EXF_GETFORCING( myTime, myIter, myThid )
+         DO j = 1,sNy
+          DO i = 1,sNx
+           SSTtmp = sstExtrapol
+     &          *( surfLevTemp(i,j,bi,bj)-theta(i,j,kl,bi,bj) )
```
It would be better to do the difference between 2 "theta", "theta(ks)" and "theta(kl)".

jm-c replied:
Few hings to consider here:
1. This `sstExtrapol` is a rather simple/simplistic way to account for strong temperature gradient near the surface. If we were to improve it, we would consider surface net heat-flux and stratification just below the surface level. And in this case we would extrapolate both conservative temp. and absolute salinity and get a new estimation of SST from these interpolated value.
2. in this PR current form, assuming I use TEOS10, the extrapolation  is applied between conservative temp at level 2 and in-situ temp. at level 1 (the name "SSTtmp" might be miss-leading, it's rather a "delta" SST here). So in places where salinity is far from mean/reference value, which translates to significant difference between conservative temp. and potential temp., this current extrapolation will result in some adjustment ("SSTtmp <> 0") even if there is not vertical temperature gradient. So I think that this is not right.
3. if instead we do the difference between 2 "theta", "theta(ks)" and "theta(kl)" (as mentioned earlier) to compute "SSTtmp" but still add this to `surfLevTemp` to get `exf_Tsf`, we don't have the same problem as above ; this still ignores salinity differences between first  and second level (but this is a general limitation of this `sstExtrapol`); we skip a second conversion form conservative Temp to in-situ Temp. but the error here is (likely very) small since the temp adjustment "SSTtmp" is likely not large and we would do this conversion using the same salinity.

mjlosch replied:
OK, makes sense. I'll do as suggested (point 3)

jm-c commented:
@mjlosch It seems that OpenAD tests are not happy. Was it the case also before this round of changes ?

mjlosch commented:
> It seems that OpenAD tests are not happy. Was it the case also before this round of changes ?

No, that's a new problem:
```
ERROR: SymtabError at line 0:enterType: name clash because the declaration for this symbol conflicts with an earlier declaration using the same name
	for entry[SymtabEntry(22623690036304) "TPN00" -> entryKind=unknown, typetab_id=8, origin=module:EOS_mod]
make[1]: *** [ad_input_code_sf.pre.f90] Error 1
```

The `TPN`-coefficient series was local to `S/R CONVERT_CT2PT` (with parameter statements) and I moved this to `EOS.h` and `ini_eos.F`. I can try to move it back, but maybe there's something else I overlooked?

mjlosch commented:
> but maybe there's something else I overlooked?

of course there was: `pkg/openad/seawater.F`, now fixed with [dbd31cb](https://github.com/MITgcm/MITgcm/pull/812/commits/dbd31cb87e221cfed3870c8c463677e5bcf392c5).

I think I addressed all points raised by @jm-c except for (3): the name of the variable `surfLevTemp`, formerly `seaSurfTemp`. I don't have a convincing idea. The variable is (supposed to be) the potential temperature at the surface of the ocean, which is the same as the in-situ temperature for `k=1` in `z`-coordinates. Under ice shelves, the potential temperature is converted to the in-situ temperature (and the conversion for conservative temperature to potential temperature is missing so far, need to add that), so the field `surfLevTemp` still (should) contain the potential temperature, but it is no longer "SST". A name that reflects the use as the (potential) temperature at the top of the water column could be (in analogy to `kTopC`): `thetaTop` (or `topTheta`) or more explicitly `topPotTemp`. @jm-c's susggestion `gcmSST` does not imply this, maybe `gcmTPT`?

jm-c commented:
I looked a little bit at `pkg/shelfice`, and the conversion there form Pot.Temp to in-situ is using `SW_TEMP` with coeff. from 1983 Unesco EOS, and not a TEOS10 updated version (if using TEOS10). My point here is that if we were to update `pkg/shelfice` we would probably not use `surfLevTemp` but instead convert directly the Conservative Temp (CT) to an in-situ Temp at the pressure under the ice: Only do a single conversion ; Plus I would naively think that the new function here to convert CT to Pot.Temp would also be dependent on Pressure (if not at the surface), but I might be wrong ; and then there is the `SHELFICEBoundaryLayer` special case.
But on the other side, we might use  `surfLevTemp` inside a cost-function to compared with SST observations.
So at the end, since it's currently only defined at the surface level, not clear if it is a Pot.Temp or In-situ Temp.

One last thing, it would be nice to add a diagnostics for `surfLevTemp`. Again, there is the issue of how this new diag will be called (and described), but I think it would be useful to have it when using TEOS10.

And I will update the reference output soon.

mjlosch commented:
As a note: I agree, it's not quite so straightforward to include this conversion in pkg/shelfice, but pressure does not appear to be required for the conversion ct->pt, e.g. here: https://teos-10.github.io/GSW-R/reference/gsw_pt_from_CT.html
The conversion ct-> insitu temp does require pressure: https://teos-10.github.io/GSW-R/reference/gsw_t_from_CT.html

As usual, I claim that I do not care too much about the name ...

mjlosch commented:
I think that the OpenAD issues are solved. TAF seems to be happy, too (but not all tests are done).

jm-c commented:
I run the TAF adm testreport and few experiments show very small differences in AD monitor + only one with AD-grad minor difference (15 matching digits vs 16, for global_ocean.cs32x15.thsice). likely dues to the changes in TAF store dir.
Running the Tapenade test now to check.

mjlosch commented:
The small AD-test changes make sense in the context of the modified storage.

Feel free to rename `surfLevTemp` to anything you like (note that there are some instances in the comments/headers that I would like miss, given my sloppiness).

jm-c commented:
OK, I did run the Tapeande -adm testreport and the 2 lab_sea tests (primary + secondary "noecco" tests) show small differences for Ad-Grad, 13 mactching digits instead of 16 previously. I re-run to double-check and if it's confirmed, will update these 2 ref. output.

jm-c commented:
I am going to rename `surfLevTemp` (+ change description in FFIELDS.h) to just reflect what it is now (instead of what it could be in the future), since it's not very clear how/if it will be defined under ice-shelf. And if we stick to this, a "SST" something name seems better (clear that this is different for model surface temp `theta(:,:,ks)`, at least if using TEOS10), so propose to go with "gcmSST". Is this OK @mjlosch and @jrscott ?

mjlosch commented:
I think that `gcmSST` is OK.

jrscott commented:
hmmm I don't love the gcm bit, and as non-expert would have no idea what/why this variable exists (whereas surfLevTemp is at least more suggestive of what it contains, issues discussed above aside).  So it seems to avoid possible mislead/confusion it is now named something a bit inscrutable, not sure this is the best route. If you want to keep this name I'd expand the explanation in FFIELDS.h a bit further to say more about purpose/use and difference in TEOS-10 etc. (realize this may change somewhat in future...); I don't have an alternative suggestion for name however :(

jm-c commented on model/inc/FFIELDS.h:
```diff
@@ -111,6 +111,12 @@ C     EfluxP - p-component of Eliassen-Palm flux vector
       _RS  pLoad    (1-OLx:sNx+OLx,1-OLy:sNy+OLy,nSx,nSy)
       _RS  sIceLoad (1-OLx:sNx+OLx,1-OLy:sNy+OLy,nSx,nSy)
 
+C     surfLevTemp :: in-situ temperature of the top most wet level
+C                    generally this is the surface, but it can be the
+C                    temperature in the first level under an ice shelf.
```
```suggestion
C     gcmSST :: model in-situ Sea Surface Temperature (SST); corresponds to
C               surface-level model variable "theta", except if using TEOS-10 ;
C               in that case a conversion from model Conservative Temperature 
C               "theta" is applied. Note: not defined under an ice-shelf

```

jm-c replied:
This is my suggestion regarding the updated description of "gcmSST" (not longer than 80.c). @jrscott, @mjlosch does this sounds good ?

jm-c replied:
And in the mean time, will apply this renaming everywhere else.

mjlosch replied:
Looks good to me.

jrscott replied:
thanks @jm-c LGTM

jm-c commented:
@mjlosch I think it would be good to add a diagnostics for this `gcmSST` (formerly `surfLevTemp`). Do you agree ? and if yes do you want to take care of this (if you don't have time I could push something).

mjlosch commented:
@jm-c I don't see myself doing this now. Please go ahead and push your modifications.

jm-c commented:
@mjlosch I added a new diagnostics ("GCM_SST ") and also updated the main-diagnostics title to mention "Conservative Temperature" if using TEOS10 (previously was "Potential Temperature"). This update keeps the same diags title when not using TEOS10. Please take a look and if you have better suggestions (e.g., "short name" `tTitl2` could be a little bit longer, but would be better, from my point of view, to be shorter than the full "Conservative Temperature"), don't hesitate.

mjlosch commented:
@jm-c I am fine with this changes!

jm-c commented:
OK, form my side this PR is ready to be merged, and since the latest changes work for @mjlosch , could merge it soon. @jrscott do you want/need to check the latest changes there ?

jrscott commented:
@jmc I  looked through changes in past two months -- LGTM

jm-c commented:
Going to cook something for tag-index and merge this PR soon after.
